vyper/parser/context.py

```diff
@@ -71,25 +72,35 @@ def __init__(self, vars, global_ctx, sigs=None, forvars=None, return_type=None,
         # store global context
         self.global_ctx = global_ctx
 
-    def set_in_assignment(self, state: bool):
-        self.in_assignment = state
+    def is_constant(self):
+        return self.constancy == Constancy.Constant or self.in_assertion
 
-    def set_in_for_loop(self, name_of_list):
-        self.in_for_loop.add(name_of_list)
+    #
+    # Context Managers
+    # - Context managers are used to ensure proper wrapping of scopes and context states.
 
-    def remove_in_for_loop(self, name_of_list):
+    @contextmanager
+    def set_in_for_loop_scope(self, name_of_list):
+        self.in_for_loop.add(name_of_list)
+        yield
         self.in_for_loop.remove(name_of_list)
 
-    def is_constant(self):
-        return self.constancy == Constancy.Constant or self.in_assertion
+    @contextmanager
+    def set_in_assignment(self):
+        self.in_assignment = True
+        yield
+        self.in_assignment = False
 
-    def set_in_assertion(self, val):
-        self.in_assertion = val
+    @contextmanager
+    def set_in_assertion(self):
```

Comment:
Maybe these would be more semantically meaningful names if the methods were `assertion_scope`.

```python
with self.context.assertion_scope():
    ...
```

I thought of `assertion_context` as well but the duplicate use of the term context bugged me a bit.

Reply:
I agree let's call them _context scopes_ or _scopes_ for short.